modules/hal_rpi_pico/CMakeLists.txt

```diff
@@ -8,60 +10,54 @@ if(CONFIG_HAS_RPI_PICO)
 	set(common_dir ${ZEPHYR_HAL_RPI_PICO_MODULE_DIR}/src/common)
 	set(boot_stage_dir ${rp2_common_dir}/boot_stage2)
 
-	# Compile the second stage bootloader as a separate executable
-
-	add_executable(boot_stage2)
-
-	target_sources_ifdef(CONFIG_RP2_FLASH_W25Q080 boot_stage2 PRIVATE
-				${boot_stage_dir}/boot2_w25q080.S)
-	target_sources_ifdef(CONFIG_RP2_FLASH_GENERIC_03H boot_stage2 PRIVATE
-				${boot_stage_dir}/boot2_generic_03h.S)
-	target_sources_ifdef(CONFIG_RP2_FLASH_IS25LP080 boot_stage2 PRIVATE
-				${boot_stage_dir}/boot2_is25lp080.S)
-	target_sources_ifdef(CONFIG_RP2_FLASH_W25X10CL boot_stage2 PRIVATE
-				${boot_stage_dir}/boot2_w25x10cl.S)
-	target_sources_ifdef(CONFIG_RP2_FLASH_AT25SF128A boot_stage2 PRIVATE
-				${boot_stage_dir}/boot2_at25sf128a.S)
-
-	target_include_directories(boot_stage2 PUBLIC
-		${CMAKE_CURRENT_LIST_DIR}
-		${boot_stage_dir}/asminclude
-		${rp2_common_dir}/pico_platform/include
-		${rp2040_dir}/hardware_regs/include
-		${common_dir}/pico_base/include
-		${ZEPHYR_BASE}/include
-		)
-
-	target_link_options(boot_stage2 PRIVATE
-		"-nostartfiles"
-		"--specs=nosys.specs"
-		"LINKER:--script=${boot_stage_dir}/boot_stage2.ld"
-		)
-
-	# The second stage bootloader is compiled without kconfig definitions.
-	# Therefore, in order to use toolchain.h, it needs to define CONFIG_ARM.
-	target_compile_definitions(boot_stage2 PRIVATE -DCONFIG_ARM=1)
-
-	# Generates a binary file from the compiled bootloader
-	add_custom_target(bootloader_bin ALL DEPENDS boot_stage2.bin)
-	add_custom_command(OUTPUT boot_stage2.bin
-		DEPENDS boot_stage2
-		COMMAND ${CMAKE_OBJCOPY} -Obinary $<TARGET_FILE:boot_stage2> boot_stage2.bin
-		)
+	# The Second Stage Bootloader is only linked to the app that resides
+	# at 0x100. Therefore, only if the app's offset is 0x100, the second
+	# stage bootloader should be compiled.
+	if(CONFIG_RP2_REQUIRES_SECOND_STAGE_BOOT)
+		if(CONFIG_RP2_FLASH_W25Q080)
+			set(flash_type_file boot2_w25q080.S)
+		elseif(CONFIG_RP2_FLASH_GENERIC_03H)
+			set(flash_type_file boot2_generic_03h.S)
+		elseif(CONFIG_RP2_FLASH_IS25LP080)
+			set(flash_type_file boot2_is25lp080.S)
+		elseif(CONFIG_RP2_FLASH_W25X10CL)
+			set(flash_type_file boot2_w25x10cl.S)
+		elseif(CONFIG_RP2_FLASH_AT25SF128A)
+			set(flash_type_file boot2_at25sf128a.S)
+		else()
+			message(FATAL_ERROR "No flash type selected")
+		endif()
```

Comment:
This belongs in this file [bootloader/CMakeLists.txt](https://github.com/zephyrproject-rtos/zephyr/blob/1a14136228a1d62929367123251e94e2114cddd4/modules/hal_rpi_pico/bootloader/CMakeLists.txt):

What you may do here, is to pass on the flash knowledge, like this:
```
    foreach(flash W25Q080 GENERIC_03H IS25LP080 W25X10CL AT25SF128A)
      if(CONFIG_RP2_FLASH_${flash})
        set(flash_type ${flash})
        break()
      endif()
    endforeach()
```

and then pass that info to `second_stage_bootloader` as:
```
-DFLASH_TYPE=${flash_type}
```
and then translate it to proper file to use.

Bonus info: when #41301 / #40555 is in place, we can just pass the `DOTCONFIG` and easily load all Kconfig settings to the second build. But that's for another day.